### PR TITLE
[FIVE-303] (Backend) Modificar tablas Artifacts, ArtifactType, y Providers en Base de Datos

### PR DESCRIPTION
**Pull request by martinmena**
## Descripción
En base a la necesidad de guardar la información acerca de que artefacto específico se está almacenando para luego poder instanciar un objeto específico para cada uno con su cálculo del precio, se resolvió modificar la base de datos como se puede ver en la imagen (lo que se encuentra en rojo es eliminación y en verde lo nuevo). Con estas modificaciones, todo artefacto tiene asociado un ArtifactType (no se estaba usando ese campo hasta el momento), los ArtifactType tienen un nombre que representa el servicio específico, y tienen asociado un proveedor (de haberse mantenido esa columna en Artifacts no había una relación directa entre los proveedores y sus servicios o artefactos, por lo que la información podía volverse incoherente).
ChangesDB

## Tareas a realizar
- Crear la tabla **Providers** con un campo “_Id_” como clave primaria, y un campo “_Name_” para almacenar los nombres (varchar).
- En la tabla **Artifacts** se debe eliminar la columna “_Provider_”.
- En la tabla **ArtifactType** se deben agregar las columnas “_RequiredFields_” (nvarchar) y “_ProviderId_” (foreign key de la tabla Providers).

## Cambios realizados
Además de realizar los cambios en la base de datos como se específica más arriba, se realizaron los cambios necesarios para que la aplicación pudiera seguir funcionando de manera correcta. Esto significa actualizar los EntityModels, Models y Dtos de Artefactos y ArtifactTypes, además de un EntityModel, Model y Dto para trabajar con los Providers (junto con sus mapeos necesarios).

Por último, se actualizó el servicio de artefactos, y los test de dicho servicio, del servicio de presupuestos, y de el controller de artefactos. Durante las pruebas del servicio de artefactos se detectaron problemas con las respuestas del Update y Save y se corrigieron (esto se daba porque se intentaba instanciar un objeto específico en base al proveedor del artefacto, pero no se tenía esa información luego de crearlo o modificarlo).

En otra historia se encontrarán las modificaciones necesarias sobre el FrontEnd para adaptarlo a estos cambios.

### CODE REVIEW

**Review by gperelstein — APPROVED** [review]
Excellent!

**Review by matiasbeckerle — APPROVED** [review]
Changes looks good but I would like to leave a comment. I bet you have been thinking hard on this and I believe this is an important decision with no return (or hard to go to the previous stage). My question is: do you gain something by having the 3 provider names in a database table?

File: FiveRockingFingers/FRF.DataAccess/EntityModels/Provider.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace FRF.DataAccess.EntityModels
+{
+    public class Provider
+    {
+        [Key]
+        public int Id { get; set; }
+        public string Name { get; set; }
+    }
```
Comment (gonzalorom) [inline FiveRockingFingers/FRF.DataAccess/EntityModels/Provider.cs:10]: The convention for the project in a one-to-many relationship is to include navigation property at both ends. So here there is one navigator missed `public ICollection<ArtifactType> ArtifactType`. _Does it work as it is now?_ ***Yes***, is just to follow the convention.

**Review by gonzalorom — APPROVED** [review]
I like how it was implemented.

**Comment by martinmena** [comment]
> Changes looks good but I would like to leave a comment. I bet you have been thinking hard on this and I believe this is an important decision with no return (or hard to go to the previous stage). My question is: do you gain something by having the 3 provider names in a database table?

I think there is one big advantage in having the provider names in the database, and that is that every artifact has a provider associated. This means that if in the future we have 5/6 different providers, and for example you want to create a project based on one of them, this relation allows you to filter every artifact by provider, instead of searching through an endless list. The other point is that multiple provides might have services with the same name (with this one I'm just speculating) and this would be a way to identify which is which.
Now, I don't know if the providers "name" is actually needed, but some kind of relation is useful in my opinion, and the only other option that I can think of right now is that we store only a number in the ArtifactType (working as a ProviderId but without the table) and then we map that number to the correspondent provider in the BackEnd (I personally don't like this approach, it's like "simulating" a table in the backend, and not in the database).

File: FiveRockingFingers/FRF.DataAccess/EntityModels/Provider.cs (hunk 1 above)
Comment (martinmena) [reply FiveRockingFingers/FRF.DataAccess/EntityModels/Provider.cs:10]: Good point! This is working right now, but I add it to follow the same convention.

**Comment by matiasbeckerle** [comment]
The "simulation" you mention is the solution I was thinking about. I don't see it that bad as you see it 😁 -- anyways I'm OK with whatever you decide doing.

**Comment by gonzalorom** [comment]
In my opinion, I like the current approach better. I personally feel that we would be adding more lines of code, enums (or whatever class we use), which can be solved by having these values in the database 🤔.

**Pull request merged**